### PR TITLE
Using a HTML canvas for computing the text width

### PR DESCRIPTION
**Pull request by samelhusseini**
## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves

Better Flyout performance

### Proposed Changes

Use a HTML canvas injected into the document body for computing text width calculations. 
@rachel-fenichel mentioned that this method might be more performant.

This article goes into further detail of the rationale: https://observablehq.com/@sto3psl/text-measurement-in-the-browser

This method is preferred to using ``getComputedTextWidth`` and ``getBBox`` since:
- Better Flyout and Block rendering performance
- It's cross platform (including IE), whereas in the past we've had to use ``getBbox`` in IE and Edge and ``getComputedTextWidth``. 
- The SVG element no longer needs to be visible in order to calculate, ie: we can probably avoid having to re-render the entire workspace when a workspace is reset to visible. 

Across the major browsers, this change boost rendering by **~1.5x** and flyout performance by **~1.3x**.

### Reason for Changes

Performance

### Test Coverage

Tested on:
* Desktop Chrome
* Desktop Firefox
* Desktop Safari
* IOS Safari

### Additional Information

Performance analysis: 

Opening and closing the flyout 1000 times:

| Browser | Canvas  | getComputedTextWidth    | getBBox | Difference   |
|---------|---------|----------|--------------|--------------|
| Chrome  | 8827ms  |  12826ms | | 1.45 x faster |
| Firefox | 10285ms  | 13521ms  | | 1.3 x faster |
| Safari  | 14197ms  | 17738ms   | | 1.25 x faster |
| IOS Safari  | 3866ms  | 4944ms   | | 1.27 x faster |

Spaghetti test (5 times): 

| Browser | Canvas  | getComputedTextWidth    | getBBox | Difference   |
|---------|---------|----------|--------------|--------------|
| Chrome  | 10532ms  |  16147ms | | 1.5 x faster |
| Firefox | 16168ms  | 27014ms  | | 1.6 x faster |
| Safari  | 14622ms  | 30509ms   | | 2 x faster |

Airstrike test (100 times): 

| Browser | Canvas  | getComputedTextWidth    | getBBox | Difference   |
|---------|---------|----------|--------------|--------------|
| Chrome  | 7960ms  |  13808ms | | 1.73 x faster |
| Firefox | 10416ms  | 12270ms  | | 1.17 x faster |
| Safari  | 11265ms  | 15339ms   | | 1.36 x faster |

Caching of text width was removed during testing.

@NeilFraser @rachel-fenichel  FYI

### CODE REVIEW

**Comment by samelhusseini** [comment]
I'm going to hold off on pushing this till I get a chance to do more perf testing on it.

**Comment by samelhusseini** [comment]
- [ ] Try airstrike and compare results.

**Comment by samelhusseini** [comment]
@RoboErikG This is ready for another look. The new plan is to keep both the old compute method and introduce a new ``getFastTextWidth`` method. The new method requires that we know the font size and family in advance. This isn't always the case in Blockly (although it could be in the future), eg: the flyout button text.
So for most calculations (.blocklyText), I've switched to using the faster method.

The latest benchmarks put this at about 1.5x improvement in flyout performance on Chrome, and 2x improvement with render (airstrike).

File: core/utils/dom.js [hunk 1]
```diff
@@ -271,3 +278,48 @@ Blockly.utils.dom.getTextWidth = function(textElement) {
...
+    // Inject the canvas element used for computing text widths.
+    var computeCanvas = document.createElement('canvas');
+    computeCanvas.className = 'blocklyComputeCanvas';
+    document.body.appendChild(computeCanvas);
+
+    // Initialize the HTML canvas context and set the font.
+    // The context font must match blocklyText's fontsize and font-family
+    // set in CSS.
+    Blockly.utils.dom.canvasContext_ = computeCanvas.getContext('2d');
+    Blockly.utils.dom.canvasContext_.font = fontSize + 'pt ' + fontFamily;
```
Comment (RoboErikG) [inline core/utils/dom.js]: Don't you need to set the fontsize and fontfamily each time this is called since those are being passed in to the method?

Comment (samelhusseini) [reply core/utils/dom.js]: Oh yes thats true. I'll update that and run some quick tests to make sure it is still faster :)

**Comment by samelhusseini** [comment]
Latest results on Chrome: 
```
Benchmark flyout ran 1000 times old: 13207.60 new: 9256.82 1.43% faster
Benchmark airstrike ran 100 times old: 20563.90 new: 11253.72 1.83% faster
Benchmark spaghetti ran 5 times old: 11158.07 new: 8100.35 1.38% faster
```

**Comment by RoboErikG** [comment]
btw, your speed benchmark should convert the value to a % or swap the % for a × -> 1.43× faster
;)

**Comment by samelhusseini** [comment]
haha yes.

**Pull request merged**